File: services/gateway/gateway-api/src/main/java/com/example/gateway/api/gateway/AlphaVantageGateway.java
//package com.example.gateway.api.gateway;
//
//import com.example.gateway.api.model.alphavantage.GlobalQuoteResponse;
//import com.example.stockdata.api.spec.model.ShareDataResponse;
//import org.springframework.beans.factory.annotation.Value;
//import org.springframework.http.HttpEntity;
//import org.springframework.http.HttpMethod;
//import org.springframework.http.ResponseEntity;
//import org.springframework.stereotype.Service;
//import org.springframework.web.client.RestTemplate;
//import org.springframework.web.util.UriComponentsBuilder;
//
//import java.net.URI;
//
//@Service
//public class AlphaVantageGateway {
//
//    private final RestTemplate restTemplate;
//    private final String API_URL;
//    private final String API_KEY;
//    private final String URL_TEMPLATE = "query?function=GLOBAL_QUOTE&symbol=?&apikey=?&datatype=csv";
//
//    public AlphaVantageGateway(RestTemplate restTemplate,
//                               @Value("${gateway.alphavantage.url}") String API_URL,
//                               @Value("${gateway.alphavantage.key}") String API_KEY) {
//        this.restTemplate = restTemplate;
//        this.API_URL = API_URL;
//        this.API_KEY = API_KEY;
//    }
//
//    public GlobalQuoteResponse getQuoteResponse(String symbol) {
//        ResponseEntity<String> responseEntity =
//                restTemplate.getForEntity(buildUri(symbol, API_KEY), String.class);
//        String csv = responseEntity.getBody();
//        System.out.println(csv);
//
//        return new GlobalQuoteResponse.Builder()
//                .withPrice(123)
//                .withSymbol(symbol)
//                .build();
//    }
//
//    private URI buildUri(String symbol, String apiKey) {
//        UriComponentsBuilder builder = UriComponentsBuilder
//                .fromHttpUrl(API_URL)
//                .path("/query")
//                .queryParam("function", "GLOBAL_QUOTE")
//                .queryParam("symbol", symbol)
//                .queryParam("datatype", "csv")
//                .queryParam("apikey", apiKey);
//        return builder.build().toUri();
//    }
//
////symbol,open,high,low,price,volume,latestDay,previousClose,change,changePercent
////MSFT,151.0700,152.2100,150.9100,151.3600,15282906,2019-12-09,151.7500,-0.3900,-0.2570%
//}

File: services/gateway/gateway-api/src/main/kotlin/com/example/gateway/api/rest/gateway/AlphaVantageGateway.kt
package com.example.gateway.api.rest.gateway

import arrow.core.Either
import com.example.common.exception.ApiException
import com.example.common.exception.ApiExceptionCode
import com.example.gateway.api.core.model.SharePriceModel
import com.example.gateway.api.rest.converter.AlphaVantageCsvConverter
import org.springframework.beans.factory.annotation.Value
import org.springframework.http.HttpEntity
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpMethod
import org.springframework.stereotype.Service
import org.springframework.web.client.RestClientException
import org.springframework.web.client.RestTemplate
import org.springframework.web.util.UriComponentsBuilder

@Service
class AlphaVantageGateway(private val restTemplate: RestTemplate,
                          @Value("\${gateway.alphavantage.url}") private val API_URL: String,
                          @Value("\${gateway.alphavantage.key}") private val API_KEY: String) {

    private final val URL_TEMPLATE = "/query"

    fun getQuoteResponse(symbol: String): Either<ApiException, SharePriceModel> {
        val builder = UriComponentsBuilder
                .fromHttpUrl(API_URL)
                .path(URL_TEMPLATE)
                .queryParam("function", "GLOBAL_QUOTE")
                .queryParam("symbol", symbol)
                .queryParam("interval", "5min")
                .queryParam("outputsize", "full")
                .queryParam("apikey", API_KEY)
                .queryParam("datatype", "csv")


        val headers = HttpHeaders()
        headers.set("Content-Type", "application/json")

        return try {
            val responseEntity = restTemplate
                    .exchange(builder.build("").toString(), HttpMethod.GET, HttpEntity<Any>(headers), String::class.java)

            if (responseEntity.body != null) {
                AlphaVantageCsvConverter.toModel(responseEntity.body!!)
            } else {
                Either.Left(ApiException(ApiExceptionCode.EXTERNAL_API_DATA_CONVERSION_ERROR))
            }
        } catch (rce: RestClientException) {
            Either.Left(ApiException(ApiExceptionCode.EXTERNAL_API_CONNECTION_ERROR, "Could not connect to AlphaVantageApi"))
        } catch (ex: Exception) {
            Either.Left(ApiException(ApiExceptionCode.EXTERNAL_API_OTHER_ERROR, ex.message))
        }
    }
}